File: shared/schemas/target_detail.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field

from shared.enums.target import TargetType
from shared.enums.task_status import TaskStatus
from shared.models.bgp_summary import BgpSummaryRead
from shared.models.dns import DnsLookupRead
from shared.models.organization import OrganizationSummary
from shared.models.tag import TagSummary
from shared.models.whois import WhoisRecordRead


class AnnouncedPrefixDetail(BaseModel):
    prefix: str
    ip_version: int = 4
    first_seen: datetime | None = None
    last_seen: datetime | None = None


class ASNNeighbourDetail(BaseModel):
    neighbour_asn: int
    relationship: str
    power: int = 0


class ASOverviewDetail(BaseModel):
    asn: int
    holder: str = ""
    rir: str | None = None
    announced: bool = False
    block_name: str | None = None
    block_resource: str | None = None


class NetworkInfoDetail(BaseModel):
    ip: str
    prefix: str = ""
    asn: int = 0


class PrefixOverviewDetail(BaseModel):
    prefix: str
    asn: int
    holder: str = ""
    is_announced: bool = True


class RelatedPrefixDetail(BaseModel):
    related_prefix: str
    relationship: str
    origin_asn: int | None = None


class AbuseContactDetail(BaseModel):
    resource: str
    abuse_email: str
    rir: str | None = None


class TargetBgpDetailResponse(BaseModel):
    """Full BGP enrichment data for a target."""

    target_id: uuid.UUID
    target_type: TargetType
    status: TaskStatus
    summary: BgpSummaryRead | None = None
    as_overview: ASOverviewDetail | None = None
    announced_prefixes: list[AnnouncedPrefixDetail] = Field(default_factory=list)
    neighbours: list[ASNNeighbourDetail] = Field(default_factory=list)
    network_info: list[NetworkInfoDetail] = Field(default_factory=list)
    prefix_overview: list[PrefixOverviewDetail] = Field(default_factory=list)
    related_prefixes: list[RelatedPrefixDetail] = Field(default_factory=list)
    abuse_contacts: list[AbuseContactDetail] = Field(default_factory=list)


class TargetDnsDetailResponse(BaseModel):
    target_id: uuid.UUID
    target_type: TargetType
    status: TaskStatus
    error: str | None = None
    lookup: DnsLookupRead | None = None


class TargetWhoisDetailResponse(BaseModel):
    target_id: uuid.UUID
    target_type: TargetType
    status: TaskStatus
    error: str | None = None
    record: WhoisRecordRead | None = None


class TargetDetailRead(BaseModel):
    id: uuid.UUID
    target_value: str
    display_name: str | None = None
    target_type: TargetType
    project_id: uuid.UUID
    created_at: datetime
    updated_at: datetime
    created_by: uuid.UUID
    organizations: list[OrganizationSummary] = Field(default_factory=list)
    tags: list[TagSummary] = Field(default_factory=list)

    whois_status: TaskStatus
    whois_error: str | None = None
    whois: WhoisRecordRead | None = None

    dns_status: TaskStatus
    dns_error: str | None = None
    dns: DnsLookupRead | None = None

    bgp_status: TaskStatus
    bgp: TargetBgpDetailResponse | None = None


class EnrichmentRefreshResponse(BaseModel):
    target_id: uuid.UUID
    enrichment_type: str
    status: str
    message: str

File: frontend/src/lib/types/target-detail.ts
import type { TaskStatus } from './task-status';
import type { TargetType, OrganizationSummary, TagSummary } from './target';
import type { WhoisRecordRead } from './whois';

export interface DnsRecordRead {
	id: string;
	record_type: string;
	value: string;
	priority: number | null;
	weight: number | null;
	port: number | null;
	soa_email: string | null;
	soa_serial: number | null;
	caa_tag: string | null;
	caa_flag: number | null;
}

export interface DnsLookupRead {
	id: string;
	host: string;
	status_code: string;
	cdn: boolean;
	cdn_name: string;
	queried_at: string;
	record_counts: Record<string, number>;
	records: DnsRecordRead[];
}

export interface AnnouncedPrefixDetail {
	prefix: string;
	ip_version: number;
	first_seen: string | null;
	last_seen: string | null;
}

export interface ASNNeighbourDetail {
	neighbour_asn: number;
	relationship: string;
	power: number;
}

export interface ASOverviewDetail {
	asn: number;
	holder: string;
	rir: string | null;
	announced: boolean;
	block_name: string | null;
	block_resource: string | null;
}

export interface NetworkInfoDetail {
	ip: string;
	prefix: string;
	asn: number;
}

export interface PrefixOverviewDetail {
	prefix: string;
	asn: number;
	holder: string;
	is_announced: boolean;
}

export interface RelatedPrefixDetail {
	related_prefix: string;
	relationship: string;
	origin_asn: number | null;
}

export interface AbuseContactDetail {
	resource: string;
	abuse_email: string;
	rir: string | null;
}

export interface TargetBgpDetailResponse {
	target_id: string;
	target_type: TargetType;
	status: TaskStatus;
	summary: unknown;
	as_overview: ASOverviewDetail | null;
	announced_prefixes: AnnouncedPrefixDetail[];
	neighbours: ASNNeighbourDetail[];
	network_info: NetworkInfoDetail[];
	prefix_overview: PrefixOverviewDetail[];
	related_prefixes: RelatedPrefixDetail[];
	abuse_contacts: AbuseContactDetail[];
}

export interface TargetDetailRead {
	id: string;
	target_value: string;
	display_name: string | null;
	target_type: TargetType;
	project_id: string;
	created_at: string;
	updated_at: string;
	created_by: string;
	organizations: OrganizationSummary[];
	tags: TagSummary[];

	whois_status: TaskStatus;
	whois_error: string | null;
	whois: WhoisRecordRead | null;

	dns_status: TaskStatus;
	dns_error: string | null;
	dns: DnsLookupRead | null;

	bgp_status: TaskStatus;
	bgp: TargetBgpDetailResponse | null;
}
